File: scripts/extract-seed.js
/**
 * extract-seed.js
 *
 * Reads your live OnTrack database and prints TypeScript INSERT OR IGNORE
 * statements for all tasks and documents, ready to paste into seed.ts.
 *
 * Usage (run from the app folder):
 *   node scripts/extract-seed.js
 *
 * The script auto-detects the database path from the default Electron
 * userData location. If your app uses a custom data path you can override:
 *   ONTRACK_DB=C:\path\to\database.db node scripts/extract-seed.js
 */

const path   = require('path')
const os     = require('os')
const fs     = require('fs')

// ── Locate the database ───────────────────────────────────────────────────────

function getDefaultDbPath() {
  const platform = process.platform
  let userData

  if (platform === 'win32') {
    userData = process.env.APPDATA
      ? path.join(process.env.APPDATA, 'ontrack-dispatch-dashboard')
      : path.join(os.homedir(), 'AppData', 'Roaming', 'ontrack-dispatch-dashboard')
  } else if (platform === 'darwin') {
    userData = path.join(os.homedir(), 'Library', 'Application Support', 'ontrack-dispatch-dashboard')
  } else {
    userData = path.join(os.homedir(), '.config', 'ontrack-dispatch-dashboard')
  }

  return path.join(userData, 'OnTrackDashboard', 'database.db')
}

const dbPath = process.env.ONTRACK_DB || getDefaultDbPath()

if (!fs.existsSync(dbPath)) {
  console.error('Database not found at:', dbPath)
  console.error('Run the app at least once to create it, or set ONTRACK_DB=/path/to/database.db')
  process.exit(1)
}

console.error('Reading database from:', dbPath)

// ── Open the database ─────────────────────────────────────────────────────────

let Database
try {
  Database = require('better-sqlite3')
} catch {
  console.error('better-sqlite3 not found. Run: npm install  (from the app folder)')
  process.exit(1)
}

const db = new Database(dbPath, { readonly: true })

// ── Escape helpers ────────────────────────────────────────────────────────────

function esc(val) {
  if (val === null || val === undefined) return 'null'
  if (typeof val === 'number') return String(val)
  // Escape single quotes and backslashes for a JS template literal string
  return "'" + String(val).replace(/\\/g, '\\\\').replace(/'/g, "\\'") + "'"
}

function escStr(val) {
  if (val === null || val === undefined) return 'null'
  // Escape for embedding inside single quotes in a JS string
  return String(val).replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/\n/g, '\\n')
}

// ── Tasks ─────────────────────────────────────────────────────────────────────

const tasks = db.prepare(
  'SELECT id, title, category, priority, due_date, time_of_day, recurring, status, notes ' +
  'FROM tasks ORDER BY id'
).all()

console.log('// ── TASKS (extracted from live database) ──────────────────────────────────')
console.log('function seedTasks(db: Database.Database): void {')
console.log('  const ins = db.prepare(')
console.log("    'INSERT OR IGNORE INTO tasks (id, title, category, priority, due_date, time_of_day, recurring, status, notes)' +")
console.log("    ' VALUES (?,?,?,?,?,?,?,?,?)'")
console.log('  )')
console.log()

for (const t of tasks) {
  const notes = t.notes ? escStr(t.notes) : null
  const notesArg = notes !== null ? `'${notes}'` : 'null'
  const args = [
    t.id,
    `'${escStr(t.title)}'`,
    `'${escStr(t.category)}'`,
    `'${escStr(t.priority)}'`,
    t.due_date ? `'${escStr(t.due_date)}'` : 'null',
    t.time_of_day ? `'${escStr(t.time_of_day)}'` : 'null',
    t.recurring,
    `'${escStr(t.status)}'`,
    notesArg,
  ].join(',')
  console.log(`  ins.run(${args})`)
}

console.log('}')
console.log()

// ── Documents ─────────────────────────────────────────────────────────────────

const docs = db.prepare(
  'SELECT id, title, category, content, driver_id, doc_type, expiry_date ' +
  'FROM documents ORDER BY id'
).all()

console.log('// ── DOCUMENTS (extracted from live database) ──────────────────────────────')
console.log('function seedDocuments(db: Database.Database): void {')
console.log('  const ins = db.prepare(')
console.log("    'INSERT OR IGNORE INTO documents (id, title, category, content, driver_id, doc_type, expiry_date)' +")
console.log("    ' VALUES (?,?,?,?,?,?,?)'")
console.log('  )')
console.log()

for (const d of docs) {
  const content  = d.content  ? escStr(d.content)  : null
  const docType  = d.doc_type ? escStr(d.doc_type)  : null
  const expiry   = d.expiry_date ? escStr(d.expiry_date) : null
  const args = [
    d.id,
    `'${escStr(d.title)}'`,
    `'${escStr(d.category)}'`,
    content  !== null ? `'${content}'`  : 'null',
    d.driver_id !== null && d.driver_id !== undefined ? d.driver_id : 'null',
    docType  !== null ? `'${docType}'`  : 'null',
    expiry   !== null ? `'${expiry}'`   : 'null',
  ].join(',')
  console.log(`  ins.run(${args})`)
}

console.log('}')

db.close()
console.error('\nDone. Paste the output above into seed.ts, replacing the existing seedTasks() and seedDocuments() functions.')

File: scripts/extract-seed.py
"""
extract-seed.py
Reads the live OnTrack database and prints TypeScript INSERT OR IGNORE
statements for all tasks and documents, ready to paste into seed.ts.

Usage (run from the app folder):
    python scripts/extract-seed.py

Override the DB path if needed:
    python scripts/extract-seed.py "C:\\path\\to\\database.db"
"""

import sqlite3
import sys
import os

# ── Locate the database ───────────────────────────────────────────────────────

def default_db_path():
    app_name = "ontrack-dispatch-dashboard"
    if sys.platform == "win32":
        base = os.environ.get("APPDATA", os.path.expanduser("~\\AppData\\Roaming"))
        return os.path.join(base, app_name, "OnTrackDashboard", "database.db")
    elif sys.platform == "darwin":
        return os.path.expanduser(f"~/Library/Application Support/{app_name}/OnTrackDashboard/database.db")
    else:
        return os.path.expanduser(f"~/.config/{app_name}/OnTrackDashboard/database.db")

db_path = sys.argv[1] if len(sys.argv) > 1 else default_db_path()

if not os.path.exists(db_path):
    print(f"ERROR: Database not found at:\n  {db_path}", file=sys.stderr)
    print("Pass the path as an argument: python scripts/extract-seed.py \"C:\\path\\to\\database.db\"", file=sys.stderr)
    sys.exit(1)

print(f"Reading from: {db_path}", file=sys.stderr)

con = sqlite3.connect(db_path)
con.row_factory = sqlite3.Row

# ── Escape helpers ────────────────────────────────────────────────────────────

def ts(val):
    """Format a Python value as a TypeScript argument for .run()"""
    if val is None:
        return "null"
    if isinstance(val, int):
        return str(val)
    # Escape backslashes and single quotes for embedding in a JS single-quoted string
    s = str(val)
    s = s.replace("\\", "\\\\")
    s = s.replace("'", "\\'")
    s = s.replace("\n", "\\n")
    s = s.replace("\r", "")
    return f"'{s}'"

# ── Tasks ─────────────────────────────────────────────────────────────────────

tasks = con.execute(
    "SELECT id, title, category, priority, due_date, time_of_day, recurring, status, notes "
    "FROM tasks ORDER BY id"
).fetchall()

print("// ── TASKS (extracted from live database) ──────────────────────────────────")
print("function seedTasks(db: Database.Database): void {")
print("  const ins = db.prepare(")
print("    'INSERT OR IGNORE INTO tasks (id, title, category, priority, due_date, time_of_day, recurring, status, notes)' +")
print("    ' VALUES (?,?,?,?,?,?,?,?,?)'")
print("  )")
print()

for t in tasks:
    args = ",".join([
        ts(t["id"]),
        ts(t["title"]),
        ts(t["category"]),
        ts(t["priority"]),
        ts(t["due_date"]),
        ts(t["time_of_day"]),
        ts(t["recurring"]),
        ts(t["status"]),
        ts(t["notes"]),
    ])
    print(f"  ins.run({args})")

print("}")
print()

# ── Documents ─────────────────────────────────────────────────────────────────

docs = con.execute(
    "SELECT id, title, category, content, driver_id, doc_type, expiry_date "
    "FROM documents ORDER BY id"
).fetchall()

print("// ── DOCUMENTS (extracted from live database) ──────────────────────────────")
print("function seedDocuments(db: Database.Database): void {")
print("  const ins = db.prepare(")
print("    'INSERT OR IGNORE INTO documents (id, title, category, content, driver_id, doc_type, expiry_date)' +")
print("    ' VALUES (?,?,?,?,?,?,?)'")
print("  )")
print()

for d in docs:
    args = ",".join([
        ts(d["id"]),
        ts(d["title"]),
        ts(d["category"]),
        ts(d["content"]),
        ts(d["driver_id"]),
        ts(d["doc_type"]),
        ts(d["expiry_date"]),
    ])
    print(f"  ins.run({args})")

print("}")

con.close()
print("\nDone. Paste the output above into seed.ts.", file=sys.stderr)
